storage/innobase/include/trx0trx.h

```diff
@@ -601,7 +601,7 @@ struct trx_t : ilist_node<>
   /** The largest encountered transaction identifier for which no
   transaction was observed to be active. This is a cache to speed up
   trx_sys_t::find_same_or_older(). */
-  trx_id_t max_inactive_id;
+  Atomic_relaxed<trx_id_t> max_inactive_id;
```

Comment:
Can you please help me understand how this atomic variable in trx_t is initialized ?

Reply:
It is a bit obfuscated. In `storage/innobase/include/ut0pool.h` you will find that `Pool::Pool()` will invoke `aligned_malloc()` and `memset_aligned` to zero-initialize the memory (4 MiB). There is a partial initialization of some fields in `Factory::init()` a.k.a. `TrxFactory::init()`. The `max_inactive_id` does not need to be one of them; in fact, we actually benefit from having this field cached in a previously used instance of `trx_t`.

It looks like `trx_create()` would dereference a null pointer in case we have more concurrent transactions than `MAX_TRX_BLOCK_SIZE / sizeof(trx_t)`. According to a comment in `Deadlock::find_cycle()`, the maximum number of concurrent transactions is `innodb_page_size / 16 * 128`, which would be at most 524,288 when using `innodb_page_size=64k`. It looks like we’d only have to exceed 381 concurrently active transactions to hit that condition, if `sizeof(trx_t)` is 2752.

Reply:
ok, memset/zero initialization is good strategy for POD. I think it is safer to initialize atomic variables explicitly. Can we do that in trx_init() ?

Reply:
Yesterday, I missed the fact that `trx_pools` uses one more class template. `trx_create()` is actually invoking `PoolManager::get()`, which may create additional `Pool` or wait for a transaction object to become available.

If we zero-initialized `trx_t::max_inactive_id`, we would reintroduce a performace regression that this field was designed to prevent in 4105017a5832c3486002c4ec0c67005ceb5ab88b ([MDEV-30357](https://jira.mariadb.org/browse/MDEV-30357)).

As far as I understand, [relaxed memory ordering](https://en.cppreference.com/w/cpp/atomic/memory_order#Relaxed_ordering) for loads and stores is [typically implemented](https://www.cl.cam.ac.uk/~pes20/cpp/cpp0xmappings.html) mostly as a compiler barrier, with no difference whatsoever to non-atomic loads or stores. For anything stronger than `std::memory_order_relaxed`, special instructions are needed, more so on ISA implementations that do not follow Total Store Ordering. On ARM before LSE and on POWER that construct would be a loop around load-acquire/store-conditional. But the relaxed loads and stores are just like regular loads and stores.

Only on 32-bit targets there could be a difference, because we are dealing with a 64-bit quantity.

In fact, it might be better to use an anonymous `union` between the non-atomic field and the `Atomic_relaxed` one in order to avoid introducing a regression on 32-bit systems. We only need the load and the store in `row_purge_check()` to be atomic; `lock_sec_rec_some_has_impl()` had better use regular access, because the `trx_t` there is almost exclusively accessed by the current thread. (The main exception is the lock conversion from implicit to explicit by other transactions that are registering a conflicting lock.)

On 32-bit processors the 64-bit atomic loads and stores are special; using regular instructions could theoretically result in torn reads. For example, if `trx_t::max_inactive_id` was being concurrently modified from 0x1ffffffff to 0x200000001, a concurrent reader might read it as 0x100000001 or 0x2ffffffff. However, I don’t think that this is likely in practice, because `trx_t` in  `trx_pools` is aligned to the cache line size, which is at least 512 bits (64 bytes). Even without that, on 32-bit systems the minimum alignment of structs should be 64 bits, for compatibility with the IEEE 754 `double` precision 64-bit floating point type. Okay, if some code was interrupted between two 32-bit loads and the context switch was long enough, then we could get a torn read.

On IA-32, I would expect a 80387 floating point register to come to the rescue, but on https://godbolt.org I checked that it’s actually using an XMM register of the Pentium MMX vintage. That makes sense, for years already, we depend on `lock cmpxchg8b`, which was introduced in the Pentium. In fact, our "x86" is `-march=i686`. There might not be a 64-bit 80387 load or store either; the register file width is 80 bits.
```c++
#include <atomic>
uint64_t f(std::atomic<uint64_t> &num)
{ return num.load(std::memory_order_relaxed); }
```
On ARMv2 through ARMv7, the atomic load would invoke a runtime library function `__atomic_load_8()`. I tested the above program on an `armhf` target, and it turns out that such a function is not available unless I specify `-latomic`. In `libatomic`. That seems to be a rather complex function, because it is also invoking `std::operator&(std::memory_order, std::__memory_order_modifier)` on an `armhf` system that I checked. On 32-bit systems other than `i686` we might in fact want to just use the `trx_t::mutex` instead of using atomic memory access here.

I will revise this accordingly.

Reply:
19acb0257eec0956e3a0dd3b11bbb701fa7ea648 adds source code comments about the initialization and the reuse of this cached value across transactions. I checked the generated code on all 32-bit and 64-bit targets I could find on https://godbolt.org. Most 32-bit targets would invoke a library function `__atomic_load_8()` for an atomic 64-bit load. The Super-H would emit a loop around two 32-bit loads, and `--march=i686` would use a 64-bit load via an XMM register instead of two regular 32-bit loads. On all 64-bit targets, the relaxed and regular loads used the same machine instructions.